Blog: Introducing HouseWatch (#6114)
PostHog/posthog.com: contents/blog/introducing-housewatch.md

---
date: 2023-06-12
title: "Introducing HouseWatch: An open source toolkit for ClickHouse"
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
category: PostHog news
author:
  - ian-vanagas
  - yakko-majuri
tags:
  - Product updates
  - Release notes
  - ClickHouse
featuredImage: ../images/blog/posthog-engineering-blog.png
featuredImageType: full
---

We are big fans of ClickHouse. We rely on it heavily to store and retrieve the massive amount of data we process every day. In doing this at scale for multiple years now, we’ve built a lot of expertise and systems related to ClickHouse. 

To formalize and share these, we’ve recently built and launched [HouseWatch](https://github.com/PostHog/HouseWatch), an open-source suite of tools for monitoring and managing ClickHouse. HouseWatch is free and works with your existing ClickHouse instance. You can [clone it from GitHub](https://github.com/PostHog/HouseWatch) and deploy it via Docker Compose.

## Why we built HouseWatch

> We started using ClickHouse in August 2021 when we [moved away from Postgres](/blog/how-we-turned-clickhouse-into-our-eventmansion). During this time, we’ve gained a lot of expertise in using and debugging ClickHouse. 

Reviewer: ```suggestion
We started using ClickHouse in August 2021 when we [moved away from Postgres](/blog/how-we-turned-clickhouse-into-our-eventmansion).
```

---
date: 2023-06-12
title: "Introducing HouseWatch: An open source toolkit for ClickHouse"
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
category: PostHog news
author:
  - ian-vanagas
  - yakko-majuri
tags:
  - Product updates
  - Release notes
  - ClickHouse
featuredImage: ../images/blog/posthog-engineering-blog.png
featuredImageType: full
---

We are big fans of ClickHouse, we rely on it heavily to store and retrieve the massive amount of data we process every day. In doing this at scale for multiple years now, we’ve built a lot of expertise and systems related to ClickHouse. 

To formalize and share these, we’ve recently built and launched [HouseWatch](https://github.com/PostHog/HouseWatch), an open-source suite of tools for monitoring and managing ClickHouse. HouseWatch is free and works with your existing ClickHouse instance. You can [clone it from GitHub](https://github.com/PostHog/HouseWatch) and deploy it via Docker Compose.

## Why we built HouseWatch

> We have used ClickHouse at PostHog since August 2021, when we [moved from Postgres](/blog/how-we-turned-clickhouse-into-our-eventmansion). During this time, we’ve gained a lot of expertise using and debugging ClickHouse. 

Reviewer 2: ```suggestion
We started using ClickHouse in August 2021 when we [moved away from Postgres](/blog/how-we-turned-clickhouse-into-our-eventmansion). During this time, we’ve gained a lot of expertise in using and debugging ClickHouse. 
```

> ClickHouse provides tons of easily queryable metadata about the system, but knowing how and what to query is difficult. From our usage of ClickHouse, we’ve built an intuition for this, some of which we’ve documented in our [ClickHouse manual](/handbook/engineering/clickhouse). 

Reviewer 2: ```suggestion
ClickHouse provides tons of easily queryable metadata about your system, but knowing how and what to query is difficult. From our usage of ClickHouse, we’ve built an intuition for this – some of which we’ve documented in our [ClickHouse manual](/handbook/engineering/clickhouse). 
```

> We’ve also built a lot of systems and processes for managing clusters. These include:

Reviewer 2: ```suggestion
We’ve also built many systems and processes for managing clusters. These include:
```

- Tracking metrics via Grafana
- Querying via Metabase
- Running operations on nodes through `ssh`
- Managing async migrations with [custom-built tools](/blog/async-migrations)

To share our expertise, formalize these tools, and centralize them in one place, Li, Yakko, and Tim built the first version of HouseWatch at our [Aruba offsite](/blog/aruba-hackathon). We felt it would be useful to us, and support other ClickHouse users as well.

## Features

> HouseWatch provides multiple battle-tested tools for monitoring and managing ClickHouse.

Reviewer 3: not sure I'd call them battle-tested as HouseWatch is still Beta and we ourselves aren't using it in prod yet.

I get what you meant with this though, which is that our experience and knowledge is battle-tested, but doesn't sound like that here

Author: reworded

### Query performance and analysis

> To help understand the performance of all the queries on your ClickHouse clusters, we list queries, with emphasis placed on active and slow queries. 

Reviewer 2: ```suggestion
To help understand the performance of all the queries on your ClickHouse clusters, we list queries and place an emphasis on active and slow queries. 
```

> To help understand the performance of all the queries on your ClickHouse clusters, we list queries, with emphasis placed on active and slow queries. 

Reviewer 3: something more like this might be better?

```suggestion
To help understand the performance of all the queries on your ClickHouse clusters, we provide a list of normalized queries and their performance metrics. You can sort queries in your cluster by average duration, memory usage, and % of IOPS, for example.
```

Reviewer 3: Oh lol I see it below now - but "we list queries" just sounded a bit off

Each includes metrics on average run time, calls per minute, percentage of all IOPs, total IOPs, and percentage of run time. It also includes details on the query itself and the `EXPLAIN` statement. From this, you can monitor and improve important, slow, or high-stress queries.

![Slow](../images/blog/introducing-housewatch/slow.png)

### Schema stats

> HouseWatch provides stats for all the tables for your cluster and lets you dive into the details for each of them. For each table, see columns, parts, compressed and uncompressed disk space sizes, disk usage, and more.

Reviewer 2: ```suggestion
HouseWatch provides stats for all the tables for your cluster and lets you dive into the details for each of them. For each table, you can see columns, parts, compressed and uncompressed disk space sizes, disk usage, and more.
```

![Schema](../images/blog/introducing-housewatch/table.png)

### Query editing and benchmarking

With the size of data ClickHouse stores and processes, optimizing your queries is critical. To help you do this, you can run and edit queries in HouseWatch, as well as test and compare them.

You can run queries on your ClickHouse cluster and receive results visualized in HouseWatch. For query comparisons, you can write two queries, run them, and then get metrics on performance across the duration, read bytes, CPU usage, memory usage, and more. 

![Compare](../images/blog/introducing-housewatch/compare.png)

We’ve also built a natural language query editor that uses GPT to create ClickHouse queries based on the table and instructions you provide.

![AI](../images/blog/introducing-housewatch/ai.png)

### Logs and errors

Like any good monitoring tool, HouseWatch provides access to logs and errors from ClickHouse. For both, you can search for specific ones, as well as see the number of and most recent occurrences.

![Errors](../images/blog/introducing-housewatch/errors.png)

### Operations

> Last but not least is our operations tool. This enables you to run and monitor long-running operations like migrations or SQL commands. You can monitor their status, pause or stop them, and retry them if they fail. Failures come with automatic rollbacks as well.

Reviewer 3: I'd mention here again that this is inspired by our async migrations tool which we've been using in production for over a year.

Maybe on one of them we link to the docs and the other to the blog? Or one can be without a link

Author: Got it

![Operations](../images/blog/introducing-housewatch/operations.png)

## Deploying HouseWatch

First, clone the repo.

```bash
git clone https://github.com/PostHog/HouseWatch
```

Next, create a `.env` file and add the following environment variables.

```
CLICKHOUSE_HOST=localhost \
CLICKHOUSE_CLUSTER=mycluster \
CLICKHOUSE_USER=default \
CLICKHOUSE_PASSWORD=xxxxxxxxxxx \
```

Finally, run Docker Compose.

```bash
docker compose -f docker-compose.yml up
```

## Future plans

We aspire for HouseWatch to be like [pganalyze](https://pganalyze.com/) for ClickHouse. There is more to build to make this a reality including:

> - System issues tab
> - `EXPLAIN` visualizer
> - Multiple instance support
> - Stats on page cache hit percentage
> - More operations controls like view, delete, edit, re-run, and display errors

Reviewer 3: ```suggestion
- An index advisor tool
- A visualizer for `EXPLAIN` statements
- Support for monitoring multiple instances
- Automatic surfacing of known system issues 
```
